core/cache/application.go

```diff
@@ -26,20 +30,41 @@ type Application struct {
 
 	details    ApplicationChange
 	configHash string
+	hashCache  *hashCache
+}
+
+// Config returns the current application config.
+func (a *Application) Config() map[string]interface{} {
+	a.mu.Lock()
+	a.metrics.ApplicationConfigReads.Inc()
+	a.mu.Unlock()
+	return a.details.Config
 }
 
-func (m *Application) setDetails(details ApplicationChange) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.details = details
+// WatchConfig creates a watcher for the application config.
+func (a *Application) WatchConfig(keys ...string) *ConfigWatcher {
+	return newConfigWatcher(keys, a.hashCache, a.hub, a.topic(applicationConfigChange))
+}
 
-	configHash, err := hash(details.Config)
-	if err != nil {
-		logger.Errorf("invariant error - application config should be yaml serializable and hashable, %v", err)
-		configHash = ""
-	}
-	if configHash != m.configHash {
-		m.configHash = configHash
-		// TODO: publish config change...
+func (a *Application) setDetails(details ApplicationChange) {
+	a.mu.Lock()
+
+	a.details = details
+	hashCache, configHash := newHashCache(
+		details.Config, a.metrics.ApplicationHashCacheHit, a.metrics.ApplicationHashCacheMiss)
+
+	if configHash != a.configHash {
+		a.configHash = configHash
+		a.hashCache = hashCache
+		a.hub.Publish(a.topic(applicationConfigChange), hashCache)
 	}
+
+	defer a.mu.Unlock()
```

Comment:
I think this should be a naked `Unlock` and not include a `defer`.

Reply:
Yes.